zrml/hybrid-router/src/lib.rs

```diff
@@ -0,0 +1,761 @@
+                who.clone(),
+                market_id,
+                asset,
+                amm_amount_in,
+                BalanceOf::<T>::zero(),
+            )?;
+
+            Ok((amount_in.checked_sub_res(&amm_amount_in)?, amm_trade_info))
+        }
+
+        fn handle_amm_trade(
+            tx_type: TxType,
+            who: AccountIdOf<T>,
+            market_id: MarketIdOf<T>,
+            asset: AssetOf<T>,
+            amount_in: BalanceOf<T>,
+            min_amount_out: BalanceOf<T>,
+        ) -> Result<Option<AmmTradeOf<T>>, DispatchError> {
+            match tx_type {
+                TxType::Buy => {
+                    match T::Amm::buy(who, market_id, asset, amount_in, min_amount_out) {
+                        Ok(amm_trade) => Ok(Some(amm_trade)),
+                        Err(ApiError::SoftFailure(AmmSoftFail::Numerical)) => Ok(None),
+                        Err(ApiError::HardFailure(dispatch_error)) => Err(dispatch_error),
+                    }
+                }
+                TxType::Sell => {
+                    match T::Amm::sell(who, market_id, asset, amount_in, min_amount_out) {
+                        Ok(amm_trade) => Ok(Some(amm_trade)),
+                        Err(ApiError::SoftFailure(AmmSoftFail::Numerical)) => Ok(None),
+                        Err(ApiError::HardFailure(dispatch_error)) => Err(dispatch_error),
+                    }
+                }
+            }
+        }
+
+        /// Fills the order from the order book if it exists and meets the price conditions.
+        /// If the order is partially filled, the remaining amount is returned.
+        ///
+        /// # Arguments
+        ///
+        /// * `tx_type` - The type of transaction (Buy or Sell).
+        /// * `orders` - A list of orders from the order book.
+        /// * `remaining` - The amount to be traded.
+        /// * `who` - The account ID of the user performing the transaction.
+        /// * `market_id` - The ID of the market.
+        /// * `base_asset` - The base asset of the market.
+        /// * `asset` - The asset to be traded.
+        /// * `price_limit` - The maximum or minimum price at which the trade can be executed.
+        #[allow(clippy::too_many_arguments)]
+        fn maybe_fill_orders(
+            tx_type: TxType,
+            orders: &[OrderId],
+            mut remaining: BalanceOf<T>,
+            who: &AccountIdOf<T>,
+            market_id: MarketIdOf<T>,
+            base_asset: BaseAsset,
+            asset: AssetOf<T>,
+            price_limit: BalanceOf<T>,
+        ) -> Result<OrderAmmTradesInfo<T>, DispatchError> {
+            let mut amm_trades = Vec::new();
+            let mut order_trades = Vec::new();
+            for &order_id in orders {
+                if remaining.is_zero() {
+                    break;
+                }
+
+                let order = match T::OrderBook::order(order_id) {
+                    Ok(order) => order,
+                    Err(_) => continue,
```

Comment:
This execution path appears to be untested.

Reply:
dba485133f06b42ed10fcf41a728fd6f2ccd2a0b